src/confirm.rs

```diff
@@ -34,279 +27,93 @@ where
-                        state: WaitForConfirmationsState::WaitForNextBlock,
-                        filter_stream: filter.stream(create.poll_interval).skip(create.confirmations),
-                        confirmation_check: create
-                            .confirmation_check
-                            .take()
-                            .expect("future polled after ready; qed"),
-                        confirmations: create.confirmations,
-                    };
-                    ConfirmationsState::Wait(future)
-                }
-                ConfirmationsState::Wait(ref mut wait) => return wait.poll_unpin(ctx),
-            };
-            self.state = next_state;
-        }
-    }
-}
-
 /// Should be used to wait for confirmations
-pub fn wait_for_confirmations<T, V, F>(
+pub async fn wait_for_confirmations<T, V, F>(
     eth: Eth<T>,
     eth_filter: EthFilter<T>,
     poll_interval: Duration,
     confirmations: usize,
     check: V,
-) -> Confirmations<T, V, F>
+) -> error::Result<()>
 where
     T: Transport,
-    V: ConfirmationCheck<Check = F> + Unpin,
+    V: ConfirmationCheck<Check = F>,
     F: Future<Output = error::Result<Option<U64>>>,
 {
-    Confirmations::new(eth, eth_filter, poll_interval, confirmations, check)
-}
-
-struct TransactionReceiptBlockNumber<T: Transport> {
-    future: CallFuture<Option<TransactionReceipt>, T::Out>,
-}
-
-impl<T: Transport> Future for TransactionReceiptBlockNumber<T> {
-    type Output = error::Result<Option<U64>>;
-
-    fn poll(mut self: Pin<&mut Self>, ctx: &mut Context) -> Poll<Self::Output> {
-        let receipt = ready!(self.future.poll_unpin(ctx))?;
-        Poll::Ready(Ok(receipt.and_then(|receipt| receipt.block_number)))
-    }
-}
-
-struct TransactionReceiptBlockNumberCheck<T: Transport> {
-    eth: Eth<T>,
-    hash: H256,
-}
-
-impl<T: Transport> TransactionReceiptBlockNumberCheck<T> {
-    fn new(eth: Eth<T>, hash: H256) -> Self {
-        TransactionReceiptBlockNumberCheck { eth, hash }
-    }
-}
-
-impl<T: Transport> ConfirmationCheck for TransactionReceiptBlockNumberCheck<T> {
-    type Check = TransactionReceiptBlockNumber<T>;
-
-    fn check(&self) -> Self::Check {
-        TransactionReceiptBlockNumber {
-            future: self.eth.transaction_receipt(self.hash),
+    let filter = eth_filter.create_blocks_filter().await?;
+    // TODO: The stream should have additional checks:
+    // * We should not continue calling next on a stream that has completed (has returned None). We expect this to never
+    //   happen for the blocks filter but to be safe we should handle this case for example by `fuse`ing the stream or
+    //   erroring when it does complete.
+    // * We do not handle the case where the stream returns an error which means we are wrongly counting it as a
+    //   confirmation.
+    let filter_stream = filter.stream(poll_interval).skip(confirmations);
+    futures::pin_mut!(filter_stream);
+    loop {
+        let _ = filter_stream.next().await;
```

Comment:
This bug already existed so this PR doesn't make it worse.

Reply:
I presume this behaviour was already broken before the refactor, right? Can we please log this as a separate issue and add an issue number to the `TODO` above?

Reply:
Yes, this PR doesn't change the behavior (bug was already there). I don't want to fix it in this PR. Added issue link.